### DISCUSSION TITLE
ReadonlyConfigError: Cannot change read-only config container full_key: num_classes reference_type=Optional[Dict[Union[str, Enum], Any]] object_type=dict

### DISCUSSION THREAD

**Discussion by pure-rgb** [rwightman/efficientdet-pytorch#158] [Q&A]
I read this thread and your comment. 
https://github.com/rwightman/efficientdet-pytorch/issues/154

Didn't get the point. Can you **precisely comment** on what should I change to in your new update? 

```
def get_net():
    config = get_efficientdet_config('tf_efficientdet_d5')
    net = EfficientDet(config, pretrained_backbone=False)
    checkpoint = torch.load('../input/efficientdet/efficientdet_d5-ef44aea8.pth')
    net.load_state_dict(checkpoint)
    
    config.num_classes = 10
    config.image_size = 512

    net.class_net = HeadNet(config, num_outputs=config.num_classes)
    return DetBenchTrain(net, config)

net = get_net()
```

```
---------------------------------------------------------------------------
ReadonlyConfigError                       Traceback (most recent call last)
<ipython-input-40-9a87be35eae8> in <module>
     15     return DetBenchTrain(net, config)
     16 
---> 17 net = get_net()

<ipython-input-40-9a87be35eae8> in get_net()
      8     net.load_state_dict(checkpoint)
      9 
---> 10     config.num_classes = 9
     11     config.image_size = 512
     12 
```

### I would highly appreciate your solvable comment rather than simply comment or close the issue without solving anything.

**Comment by innat** [comment]
try this, 

```
config = get_efficientdet_config('tf_efficientdet_d5')
config.image_size = [512,512]
config.norm_kwargs=dict(eps=.001, momentum=.01)

net = EfficientDet(config, pretrained_backbone=False)
checkpoint = torch.load('../input/efficientdet/efficientdet_d5-ef44aea8.pth')

net.load_state_dict(checkpoint)
net.reset_head(num_classes=1)
net.class_net = HeadNet(config, num_outputs=config.num_classes)

DetBenchTrain(net, config)
```

**Reply by Tolga-Karahan** [reply]
I'm getting size mismatch for this code. I trained my model, but couldn't use it for inference.

**Reply by Tolga-Karahan** [reply]
I managed to do with this:


config = get_efficientdet_config('tf_efficientdet_d5')
config.image_size = [512,512]
config.norm_kwargs=dict(eps=.001, momentum=.01)

net = EfficientDet(config, pretrained_backbone=False).to('cpu')
checkpoint = torch.load('../input/effdetweights2/best-checkpoint-009epoch(1).bin', map_location=torch.device('cpu'))['model_state_dict']

net.reset_head(num_classes=14)
net.load_state_dict(checkpoint, strict=False)
net.class_net = HeadNet(config, num_outputs=config.num_classes)

net = DetBenchPredict(net)

**Reply by rwightman** [reply]
@Tolga-Karahan more compactly, you can do
```
config = get_efficientdet_config('tf_efficientdet_d5')
config.image_size = [512,512]
config.norm_kwargs=dict(eps=.001, momentum=.01)

net = create_model_from_confg(
    config, bench_task='predict',
    checkpoint = '../input/effdetweights2/best-checkpoint-009epoch(1).bin',
    checkpoint_ema=True, # if you are loading a training checkpoint and want the ema version to be used
    num_clases=14,
)
```

**Reply by Tolga-Karahan** [reply]
Thank you for your suggestion. It looks better.

**Comment by rwightman** [comment]
@Suzan009 please make an effort to at least try to solve the problem yourself before filing a bug report. Alex' example is out of date by over half a year now. I have made efforts to make it even easier to change the head since then, as Alex' example missed the re-init of the weights. I made the config read only after creation to prevent people from using a bad configuration.

@innat example should work, you can also just look at the train script here to see how the model is init as you can init for a given num_classes with one command: https://github.com/rwightman/efficientdet-pytorch/blob/master/train.py#L270-L283

In any case, it is not a bug or a feature request and should be done via the discussion forum. I will close issues that aren't issues quickly as that's what I use to track priority items to fix / improve.
